File: go/mechanisms/hypercore/exact/client/scheme.go
package client

import (
	"context"
	"fmt"
	"time"

	"github.com/coinbase/x402/go/mechanisms/hypercore"
	"github.com/coinbase/x402/go/types"
)

type ExactHypercoreScheme struct {
	signer hypercore.HyperliquidSigner
}

func NewExactHypercoreScheme(signer hypercore.HyperliquidSigner) *ExactHypercoreScheme {
	return &ExactHypercoreScheme{
		signer: signer,
	}
}

func (c *ExactHypercoreScheme) Scheme() string {
	return hypercore.SchemeExact
}

func (c *ExactHypercoreScheme) CreatePaymentPayload(
	ctx context.Context,
	requirements types.PaymentRequirements,
) (types.PaymentPayload, error) {
	nonce := time.Now().UnixMilli()

	config, ok := hypercore.NetworkConfigs[string(requirements.Network)]
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("unsupported network: %s", requirements.Network)
	}

	isMainnet := true
	if requirements.Extra != nil {
		if val, ok := requirements.Extra["isMainnet"].(bool); ok {
			isMainnet = val
		}
	}

	hyperliquidChain := "Mainnet"
	if !isMainnet {
		hyperliquidChain = "Testnet"
	}

	amountStr, err := hypercore.FormatAmount(requirements.Amount, config.DefaultAsset.Decimals)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to format amount: %w", err)
	}

	action := hypercore.HypercoreSendAssetAction{
		Type:             "sendAsset",
		HyperliquidChain: hyperliquidChain,
		SignatureChainID: "0x3e7",
		Destination:      hypercore.NormalizeAddress(requirements.PayTo),
		SourceDex:        "spot",
		DestinationDex:   "spot",
		Token:            requirements.Asset,
		Amount:           amountStr,
		FromSubAccount:   "",
		Nonce:            nonce,
	}

	signature, err := c.signer.SignSendAsset(action)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to sign action: %w", err)
	}

	payloadMap := map[string]interface{}{
		"action":    action,
		"signature": signature,
		"nonce":     nonce,
	}

	return types.PaymentPayload{
		X402Version: 2,
		Payload:     payloadMap,
	}, nil
}

File: typescript/packages/mechanisms/hypercore/src/exact/client/scheme.ts
import type { SchemeNetworkClient, PaymentPayload } from "@x402/core/types";
import type { PaymentRequirements } from "@x402/core/types";
import type { HypercoreSendAssetAction } from "../../types.js";
import type { ClientHypercoreSigner } from "../../signer.js";
import { formatAmount } from "../../utils.js";
import { HYPERCORE_NETWORK_CONFIGS } from "../../constants.js";

type PaymentPayloadResult = Pick<PaymentPayload, "x402Version" | "payload">;

/**
 * Hypercore scheme client implementation for creating payment payloads.
 */
export class ExactHypercoreScheme implements SchemeNetworkClient {
  readonly scheme = "exact";

  /**
   * Create a client scheme with a signer.
   *
   * @param signer - Signer for SendAsset actions.
   */
  constructor(private signer: ClientHypercoreSigner) {}

  /**
   * Create a payment payload for Hypercore.
   *
   * @param x402Version - Protocol version.
   * @param requirements - Payment requirements from the server.
   * @returns Payment payload.
   */
  async createPaymentPayload(
    x402Version: number,
    requirements: PaymentRequirements,
  ): Promise<PaymentPayloadResult> {
    const nonce = Date.now();

    const config = HYPERCORE_NETWORK_CONFIGS[requirements.network];
    if (!config) {
      throw new Error(`Unsupported network: ${requirements.network}`);
    }

    const isMainnet = requirements.extra?.isMainnet !== false;

    const action: HypercoreSendAssetAction = {
      type: "sendAsset",
      hyperliquidChain: isMainnet ? "Mainnet" : "Testnet",
      signatureChainId: "0x3e7",
      destination: requirements.payTo.toLowerCase(),
      sourceDex: "spot",
      destinationDex: "spot",
      token: requirements.asset,
      amount: formatAmount(requirements.amount, config.defaultAsset.decimals),
      fromSubAccount: "",
      nonce,
    };

    const signature = await this.signer.signSendAsset(action);

    return {
      x402Version,
      payload: {
        action,
        signature,
        nonce,
      },
    };
  }
}
